src/connector/src/parser/json_parser.rs

```diff
@@ -134,11 +131,12 @@ impl JsonParser {
         if errors.is_empty() {
             Ok(())
         } else {
-            Err(RwError::from(ErrorCode::InternalError(format!(
+            // TODO(error-handling): multiple errors
+            bail!(
                 "failed to parse {} row(s) in a single json message: {}",
                 errors.len(),
-                errors.iter().join(", ")
-            ))))
+                errors.iter().format(", ")
```

Comment:
What's the difference between `join` and `format`? Is the latter a better practice? `join` looks more intuitive to me actually.

Reply:
`format` saves an unnecessary string allocation.